File: processor/internal/enrichment/gym.go
package enrichment

import (
	"github.com/pokemon/poracleng/processor/internal/geo"
	"github.com/pokemon/poracleng/processor/internal/staticmap"
)

// Gym builds enrichment fields for a gym webhook.
func (e *Enricher) Gym(lat, lon float64, teamID, oldTeamID, slotsAvailable, oldSlotsAvailable int, inBattle, ex bool, gymID string, tileMode int) (map[string]any, *staticmap.TilePending) {
	m := make(map[string]any)

	tz := geo.GetTimezone(lat, lon)
	m["conqueredTime"] = geo.FormatNow(tz, e.TimeLayout)
	m["tth"] = geo.TTH{Hours: 1}
	addSunTimes(m, lat, lon, tz)

	// Gym identity fields for DTS templates
	m["gymId"] = gymID

	// Slot and team fields for DTS templates
	m["slotsAvailable"] = slotsAvailable
	m["oldSlotsAvailable"] = oldSlotsAvailable
	m["trainerCount"] = 6 - slotsAvailable
	m["oldTrainerCount"] = 6 - oldSlotsAvailable
	m["teamId"] = teamID
	m["oldTeamId"] = oldTeamID
	m["ex"] = ex
	m["inBattle"] = inBattle

	// Icon URLs
	trainerCount := 6 - slotsAvailable
	if e.ImgUicons != nil {
		m["imgUrl"] = e.ImgUicons.GymIcon(teamID, trainerCount, inBattle, ex)
	}
	if e.ImgUiconsAlt != nil {
		m["imgUrlAlt"] = e.ImgUiconsAlt.GymIcon(teamID, trainerCount, inBattle, ex)
	}
	if e.StickerUicons != nil {
		m["stickerUrl"] = e.StickerUicons.GymIcon(teamID, trainerCount, inBattle, ex)
	}

	// Map URLs
	e.addMapURLs(m, lat, lon, "gyms", gymID)

	// Reverse geocoding
	e.addGeoResult(m, lat, lon)

	// Static map tile
	pending := e.addStaticMap(m, "gym", lat, lon, map[string]any{
		"team_id":        teamID,
		"slotsAvailable": slotsAvailable,
		"inBattle":       inBattle,
		"ex":             ex,
	}, tileMode)

	// Game data enrichment
	if e.GameData != nil {
		if info, ok := e.GameData.Util.Teams[teamID]; ok {
			m["gymColor"] = info.Color
			m["color"] = info.Color // deprecated alias
		}
	}

	e.setFallbackImg(m, e.FallbackImgGym)

	return m, pending
}

// GymTranslate adds per-language translated fields.
// lastOwnerID is the team ID of the last controller (from webhook last_owner_id).
func (e *Enricher) GymTranslate(base map[string]any, teamID, oldTeamID, lastOwnerID int, lang string) map[string]any {
	if e.GameData == nil || e.Translations == nil {
		return nil
	}

	m := make(map[string]any, 8) // only translated fields; caller merges base + perLang

	tr := e.Translations.For(lang)
	addTeamFields(m, e.GameData, tr, teamID)
	if oldTeamID >= 0 {
		if info, ok := e.GameData.Util.Teams[oldTeamID]; ok {
			m["oldTeamName"] = tr.T(info.Name)
			m["oldTeamNameEng"] = info.Name
			m["oldTeamEmojiKey"] = info.Emoji
		}
	}
	// Previous controller's team (from webhook last_owner_id — may differ from old_team_id)
	m["previousControlId"] = lastOwnerID
	if lastOwnerID >= 0 {
		if info, ok := e.GameData.Util.Teams[lastOwnerID]; ok {
			m["previousControlName"] = tr.T(info.Name)
			m["previousControlNameEng"] = info.Name
		}
	}

	return m
}

File: alerter/src/controllers/gym.js
const Controller = require('./controller')
const nightTime = require('./common/nightTime')

/**
 * Controller for processing gym webhooks
 * Alerts on Team change
 */
class Gym extends Controller {
	async handleMatched(obj, matchedUsers, matchedAreas) {
		const data = obj
		const minTth = this.config.general.alertMinimumTime || 0

		try {
			const logReference = data.id || data.gym_id

			Object.assign(data, this.config.general.dtsDictionary)
			data.gymId = data.id || data.gym_id
			if (data.gym_name) data.name = data.gym_name
			data.name = this.escapeJsonString(data.name)
			data.gymName = data.name
			data.gymUrl = data.url

			// tth and conqueredTime provided by processor enrichment
			data.applemap = data.appleMapUrl // deprecated
			data.mapurl = data.googleMapUrl // deprecated
			data.distime = data.disappearTime // deprecated

			// Stop handling if it already disappeared or is about to go away
			if (data.tth.firstDateWasLater || ((data.tth.hours * 3600) + (data.tth.minutes * 60) + data.tth.seconds) < minTth) {
				this.log.debug(`${data.id} [matched] Gym already disappeared or is about to go away in: ${data.tth.hours}:${data.tth.minutes}:${data.tth.seconds}`)
				return []
			}

			// Use processor-provided matched areas
			data.matchedAreas = matchedAreas || []
			data.matched = data.matchedAreas.map((x) => (x.name || x).toLowerCase())

			data.teamId = data.team_id ?? data.team ?? 0
			data.oldTeamId = data.old_team_id ?? 0
			data.previousControlId = data.last_owner_id ?? 0
			data.slotsAvailable = data.slots_available
			data.oldSlotsAvailable = data.old_slots_available
			data.trainerCount = 6 - data.slotsAvailable
			data.oldTrainerCount = 6 - data.oldSlotsAvailable
			data.ex = !!(data.ex_raid_eligible ?? data.is_ex_raid_eligible)
			data.inBattle = data.is_in_battle ?? data.in_battle

			const whoCares = matchedUsers

			if (whoCares.length) {
				this.log.info(`${logReference}: [matched] Gym ${data.name} and ${whoCares.length} humans cared.`)
			} else {
				return []
			}

			try {
				// Icon URLs are pre-computed by the processor enrichment

				const geoResult = await this.getAddress({ lat: data.latitude, lon: data.longitude })
				const jobs = []

				nightTime.setNightTime(data, this.config)

				// Static map is pre-computed by the processor enrichment
				data.staticmap = data.staticMap // deprecated alias
				data.intersection = await this.obtainIntersection(data)

				for (const cares of whoCares) {
					this.log.debug(`${logReference}: [matched] Creating gym alert for ${cares.id} ${cares.name} ${cares.type} ${cares.language} ${cares.template}`, cares)

					this.log.verbose(`${logReference}: [matched] Creating gym alert for ${cares.type} ${cares.id} ${cares.name} ${cares.language} ${cares.template}`)

					const language = cares.language || this.config.general.locale
					const translator = this.translatorFactory.Translator(language)
					let [platform] = cares.type.split(':')
					if (platform === 'webhook') platform = 'discord'

					// Pre-translated names from processor per-language enrichment
					const langEnrichment = this.getLanguageEnrichment(data, language)
					data.teamName = langEnrichment.teamName
					data.gymColor = langEnrichment.teamColor
					data.color = langEnrichment.teamColor
					data.oldTeamName = langEnrichment.oldTeamName
					data.previousControlName = langEnrichment.previousControlName || ''

					// Per-platform emoji lookups (using emoji keys from enrichment)
					data.teamEmojiEng = langEnrichment.teamEmojiKey ? this.emojiLookup.lookup(langEnrichment.teamEmojiKey, platform) : ''
					data.teamEmoji = translator.translate(data.teamEmojiEng)
					data.previousControlTeamEmojiEng = langEnrichment.oldTeamEmojiKey ? this.emojiLookup.lookup(langEnrichment.oldTeamEmojiKey, platform) : ''
					data.previousControlTeamEmoji = translator.translate(data.previousControlTeamEmojiEng)

					const view = {
						...geoResult,
						...data,
						time: data.distime,
						tthh: data.tth.hours,
						tthm: data.tth.minutes,
						tths: data.tth.seconds,
						now: new Date(),
						nowISO: new Date().toISOString(),
						areas: data.matchedAreas.filter((area) => area.displayInMatches).map((area) => area.name).join(', '),
					}

					const templateType = 'gym'
					const message = await this.createMessage(logReference, templateType, platform, cares.template, language, cares.ping, view)

					const work = {
						lat: data.latitude.toString().substring(0, 8),
						lon: data.longitude.toString().substring(0, 8),
						message,
						target: cares.id,
						type: cares.type,
						name: cares.name,
						tth: data.tth,
						clean: cares.clean,
						emoji: data.emoji,
						logReference,
						language,
					}

					jobs.push(work)
				}
				return jobs
			} catch (e) {
				this.log.error(`${data.id || data.gym_id}: [matched] Can't seem to handle gym (user cares): `, e, data)
				return []
			}
		} catch (e) {
			this.log.error(`${data.id || data.gym_id}: [matched] Can't seem to handle gym: `, e, data)
		}
	}
}

module.exports = Gym
